scalding-core/src/main/scala/com/twitter/scalding/typed/FlatMappedFn.scala

```diff
@@ -20,38 +20,73 @@ import java.io.Serializable
 import com.twitter.scalding.TupleConverter
 import cascading.tuple.TupleEntry
 
-/** Closures are difficult for serialization. This class avoids that. */
-sealed trait FlatMapFn[+R] extends Function1[TupleEntry, TraversableOnce[R]]
-  with java.io.Serializable {
-
-  def filter(fn2: R => Boolean): FlatMapFn[R] =
-    FilteredFn(this, fn2)
-  def flatMap[R1](fn2: R => TraversableOnce[R1]): FlatMapFn[R1] =
-    FlatMappedFn(this, fn2)
-  def map[R1](fn2: R => R1): FlatMapFn[R1] =
-    MapFn(this, fn2)
-}
+/**
+ * This is one of 4 core, non composed operations:
+ * identity
+ * filter
+ * map
+ * flatMap
+ */
+sealed trait FlatMapping[-A, +B] extends java.io.Serializable
+object FlatMapping {
+  def filter[A](fn: A => Boolean): FlatMapping[A, A] =
+    Filter[A, A](fn, implicitly)
+
+  def filterKeys[K, V](fn: K => Boolean): FlatMapping[(K, V), (K, V)] =
+    filter { kv => fn(kv._1) }
 
-/* This is the initial way we get a FlatMapFn */
-case class Converter[R](conv: TupleConverter[R]) extends FlatMapFn[R] {
-  // make sure not to start with an Iterator to keep everything lazy
-  def apply(te: TupleEntry) = Iterator(conv(te))
+  case class Identity[A, B](ev: A =:= B) extends FlatMapping[A, B]
```

Comment:
I like keeping subclasses of a sealed trait in the companion object to avoid polluting the outer namespace with things that should be implementation details.